ql-iob-plugin/ql-iob.cc

```diff
@@ -0,0 +1,315 @@
+        std::unordered_map<std::string, const PcfParser::Constraint> constraintMap;
+        for (auto& constraint : pcfParser.getConstraints()) {
+            if (constraintMap.count(constraint.netName) != 0) {
+                log_cmd_error("The net '%s' is constrained twice!", constraint.netName.c_str());
+            }
+            constraintMap.emplace(constraint.netName, constraint);
+        }
+
+        // Read and parse pinmap CSV file
+        log("Loading pinmap CSV from '%s'...\n", a_Args[2].c_str());
+        auto pinmapParser = PinmapParser();
+        if (!pinmapParser.parse(a_Args[2])) {
+            log_cmd_error("Failed to parse the pinmap CSV file!\n");
+        }
+
+        // Build a map of pad names to entries
+        std::unordered_map<std::string, std::vector<PinmapParser::Entry>> pinmapMap;
+        for (auto& entry : pinmapParser.getEntries()) {
+            if (entry.count("name") != 0) {
+                auto& name = entry.at("name");
+
+                if (pinmapMap.count(name) == 0) {
+                    pinmapMap[name] = std::vector<PinmapParser::Entry>();
+                }
+
+                pinmapMap[name].push_back(entry);
+            }
+        }
+
+        // Check all IO cells
+        log("Processing cells...");
+        log("\n");
+        log("  type       | net        | pad        | loc      | type     | instance\n");
+        log(" ------------+------------+------------+----------+----------+-----------\n");
+        for (auto cell : topModule->cells()) {
+            auto ysCellType = RTLIL::unescape_id(cell->type); 
+
+            // Not an IO cell
+            if (ioCellTypes.count(ysCellType) == 0) {
+                continue;
+            }
+
+            log("  %-10s ", ysCellType.c_str());
+
+            std::string netName;
+            std::string padName;
+            std::string locName;
+            std::string cellType;
+
+            // Get connections to the specified port
+            const auto& ioCellType = ioCellTypes.at(ysCellType);
+            const std::string port = RTLIL::escape_id(ioCellType.port);
+            if (cell->connections().count(port)) {
+
+                // Get the sigspec of the connection
+                auto sigspec = cell->connections().at(port);
+
+                // Get the connected wire
+                for (auto sigbit : sigspec.bits()) {
+                    if (sigbit.wire != nullptr) {
+                        auto wire = sigbit.wire;
+
+                        // Has to be top level wire
+                        if (wire->port_input || wire->port_output) {
+
+                            // Check if the wire is constrained. Get pad name.
+                            std::string baseName = RTLIL::unescape_id(wire->name);
+                            std::string netNames[] = {
+                                baseName,
+                                stringf("%s[%d]", baseName.c_str(), sigbit.offset),
```

Comment:
Didn't know stringf existed, where did you find it? :)

Reply:
It's defined in `yosys.h` and implemented in `yosys.cc`: https://github.com/YosysHQ/yosys/blob/08eb0821c9c62ef21e054b6a53a980019bbfa6ca/kernel/yosys.cc#L168